### DISCUSSION TITLE
To-Dos for possible 0.2.7 release

### DISCUSSION THREAD

**Discussion by ilyaZar** [rcppsmc/rcppsmc#63] [Ideas]
Hi all!

We are obviously not in a hurry, (at least I am not if you are not!), just so we won't forget anything, and as long a the discussion from the previous release is fresh, maybe we collect some ideas for a follow up release that are reasonable, including some things that we skipped for 0.2.4.

My quick thoughts (obviously highly preliminary) are:

1. Test case for a simple linear Gaussian model, estimating the NC with all 4 resampling schemes from Alg. 3-6 and doing a Monte Carlo study comparing the distribution of the (log?)-NC-estimate to the truth (just an idea, that possible needs refinement), and to the baseline sampler class as @adamjohansen suggested. (in all cases a standard BPF)
2. Going through Alg. 6 at those lines @LeahPrice suggested.
3. Improving the documentation (in whatever manner seems reasonable)

I am not sure though, but in view of a sooner than later 0.2.5, we should probably skip a Backwards sampling step, implementing this later in 0.2.6 maybe?

Any thoughts on this are  very much appreciated!

**Comment by adamjohansen** [comment]
My personal perspective would be this: I think the sensible thing to do with what time and energy we have immediately is to consolidate what you've just spent the summer working on -- time will be relatively short with GSoC essentially over and other commitments on your time. So 1 and 3 in this list in particular are worth trying for; backward sampling is a new feature... nice to have when time permits, but let's get the essentials out of the way first and see who has time and energy.

**Comment by ilyaZar** [comment]
Update based on #67 (with detailed simulation results to follow below):

## Aim

Check the estimates of quantities that we know to be unbiased in repeated simulation
- either variance estimate by repeated runs 
- or (which I currently try to implement) cSMC algorithm estimate of the reciprocal of the normalizing constant which is unbiased if the conditioning path is obtained (in a standard linear Gaussian setting) via the Kalman filter (FFBS): comparison of 1/\hat{Z} or Z/\hat{Z}, where Z is the NC for each of the different schemes with conditioning paths drawn from the stationary distribution.

The final example should plot the distribution is similar to what is done (in a previous course I attened) here: http://www.it.uu.se/research/systems_and_control/education/2017/smc/schedule/lecture10_2x2.pdf.

## Done

1. Add boilerplate to provide an example/testing/"working proof" of conditional SMC within the main package. Files are labelled as cSMCexamples etc.
2. Add KF output to estimate "ground truth" normalizing constant (via backward sampling)
3. Add log-likelihood estimates from standard SMC with multinomial, stratified, systematic, and residual resampling
4. Add log-likelihood estimates from conditonal SMC with multinomial, stratified, systematic, and residual resampling

## Question: 

1. So far, the conceptual testing is planned to be done via the histogramm plots of repeated estimates. Is that reasonable as a start @adamjohansen or are there any other ideas (or did I get you completely wrong back then)? It's not that much work yet so we can change the direction if you fancied a different route.

2. @adamjohansen if it fine to plot $log(\hat{Z})$? I am not sure, but there may be a bias on the log-scale but not for $\hat{Z}$ directly? The results for the particular data set and the simulation study look reasonable so the bais may be small (see the simulation results I am about to add in a bit)

**Reply by eddelbuettel** [reply]
I can try to take a look at the branch, but I currently have my hands full with lecture prep for next week which I as an adjunct get to do on the weekend...

**Reply by adamjohansen** [reply]
We're coming up to the start of an academic year in the UK and consequently everything is in a state of chaos so apologies for being a bit absent.

Histograms of repeated estimates would be okay; I think boxplots -- perhaps augmented with an x for the mean and a horizontal line for the true value where known -- or similar are more standard for this kind of thing because they make comparison between different algorithms a bit easier. We certainly don't need to do anything complicated.

**Comment by ilyaZar** [comment]
Some preliminary results are for the following model:

- latent state process $x_t = 0.7 x_{t-1} + w_t$
- measurement process $y_t = x_{t} + u_t$

where $u_t, w_t \sim N(0,1)$ with time series length $T=20$ and $x_0$ initialized at the expected value of the stationary prior distribution i.e. $x_0=0$.

The following code runs a simulation study with

- number of simulation runs: `simNumber = 10000`
- number of particles for each of these simulation runs `numParticles = 1000`

```
set.seed(123)
initialX <- 0.0
myModelParams <- list(stateInit = initialX,
                      phi = 0.7,
                      varStateEvol = 1,
                      varObs = 1)
TT     <- 20
parNum <- 1000
simNum <- 10000
dat <- RcppSMC::simGaussianSSM(len = TT,
                               stateInit = 0,
                               phi = myModelParams$phi,
                               varStateEvol = myModelParams$varStateEvol,
                               varObs = myModelParams$varObs)

out <- RcppSMC::compareNCestimates(dataY = dat$measurements,
                                   trueStates = dat$states,
                                   numParticles = parNum,
                                   simNumber = simNum,
                                   modelParameters = myModelParams,
                                   plot = TRUE)
```

and produces the following figures showing:

- top row:
    - plots of data measurements + true latent states 
    - true latent states + single backward simul. path from Kalman filter FFBS (which is used as reference trajectory of the conditional BPF)
 
- middle boxplots: `simNumber` of log-likelihood estimates of standard BPF + Kalman filter log-likelihood estimate (red line)
- middle boxplots: `simNumber` of log-likelihood estimates of conditional BPF where the reference trajectory is set backwards simulation draw(s) (blue line(s) in the top right plot) + Kalman filter log-likelihood estimate (red line)

I did the histograms manually (but can add the code if you want to; essentially it uses above output `out` and does a bunch of simple calls to `hist()`.

More specifically, for the conditional BPF, I used 2 slightly different reference trajectories leading to 2 simulation studies.

## Simulation I. 

A single backward simulation pass for all 10000 simulation runs, specifically:
   - draw a single reference trajectory $x^* \sim p(x_{1:T}|y_{1:T})$ from a KF backward simulation pass
   - for `n=1, ..., n=simNum=10000` run:
       -  run standard BPF with four resampling schemes: mult, strat., syst., and residual
       -  run conditional BPF with four resampling schemes: mult, strat., syst., and residual using $x^*$ as the reference trajectory

Which leads to the following results:

![Simul_I_1e3_1e5_20](https://user-images.githubusercontent.com/36898027/172149371-63e49f83-f969-4d0b-92bf-eab35b2dacf8.png)


Notably, my guesses are
1. we see that the measurements are close to the latent states (top left figure) which benefits the KF-likelihood estimate
2. the single backward simulation pass (blue line) which is $x^*$ is quite close to the true latent states which benefits the conditional BPF 

Finally, one can visualize the SMC CLT for (log)-likelihood estimates $\hat{Z}$ for the BPF/ condtional BPF/ and the Kalman filter which estimates the "ground truth", following https://www.ida.liu.se/divisions/stima/fokurser/smc/material/lectures/lecture10.pdf (slide 12/17 and 14/17)

1. On the log-scale $log(Z_{BPF} - log(Z_{KF})$
![Sim_I_1e3_1e5_20_hists_part1](https://user-images.githubusercontent.com/36898027/172152000-cb557d32-0fd3-4b5e-9f1c-6ef68916392c.png)
![Sim_I_1e3_1e5_20_hists_part2](https://user-images.githubusercontent.com/36898027/172152011-62350218-bd19-4e51-a0e1-8a35653ff284.png)

2. As the fraction $Z_{BPF}/Z_{KF}$
![Sim_I_1e3_1e5_20_hists_part3](https://user-images.githubusercontent.com/36898027/172153483-cde5cdbe-3820-4eec-865f-5ca31f80c073.png)
![Sim_I_1e3_1e5_20_hists_part4](https://user-images.githubusercontent.com/36898027/172153494-5828f00b-f907-4ade-a8b9-ec8811b40fc7.png)



## Simulation II.

For each of the simulation runs, a separate backward simulation pass, specifically
   - for `n=1, ..., n=simNum=5000` run:
       - draw a single reference trajectory $x^*_n$ from a KF backward simulation pass
       -  run standard BPF with four resampling schemes: mult, strat., syst., and residual
       -  run conditional BPF with four resampling schemes: mult, strat., syst., and residual using $x^*_n$ as the reference trajectory


Which leads to the following results:

![Simul_II_1e3_5000_20_boxplots](https://user-images.githubusercontent.com/36898027/172211224-80efdd22-682a-46f3-b58d-9ad56046fff0.png)

1. On the log-scale $log(Z_{BPF} - log(Z_{KF})$
![Sim_II_1e3_5000_20_hists_part1](https://user-images.githubusercontent.com/36898027/172211258-44d5344f-a1a9-419c-a814-ff6eb040894a.png)
![Sim_II_1e3_5000_20_hists_part2](https://user-images.githubusercontent.com/36898027/172211267-4921fc99-2a4c-46f2-8759-9ba6a63821cf.png)

2. As the fraction $Z_{BPF}/Z_{KF}$
![Sim_II_1e3_5000_20_hists_part3](https://user-images.githubusercontent.com/36898027/172211358-9fd28093-99c2-4d78-9caf-e7973f982daa.png)
![Sim_II_1e3_5000_20_hists_part4](https://user-images.githubusercontent.com/36898027/172211372-62ce76d2-54b8-4358-be64-536b3892bf4a.png)


Notably:
- I reduced the number of simulation runs to 5000 to get the results faster, but they still look as expected I think
- the backward simulation paths are averaged to display them in the top right plot (blue line)

**Reply by ilyaZar** [reply]
To verify part of the above simulation study using `FKF` with `FKF::fkf` and `FKF::fkv` instead of the manual KF and smoother implementation as per the last commit in #67 : 

```
set.seed(123)
initialX <- 0.0
myModelParams <- list(stateInit = initialX,
                      phi = 0.7,
                      varStateEvol = 1,
                      varObs = 1)
TT     <- 20
parNum <- 1000
simNum <- 500
dat <- RcppSMC::simGaussianSSM(len = TT,
                               stateInit = 0,
                               phi = myModelParams$phi,
                               varStateEvol = myModelParams$varStateEvol,
                               varObs = myModelParams$varObs)
```

Produces:
![RplotFKF](https://user-images.githubusercontent.com/36898027/173408024-bc2a2cea-c681-4731-a71b-b66ffa435a1d.png)
